dm/master/scheduler/scheduler.go

```diff
@@ -326,6 +326,49 @@ func (s *Scheduler) AddSourceCfg(cfg *config.SourceConfig) error {
 	return nil
 }
 
+// UpdateSourceCfg update the upstream source config to the cluster.
+// please verify the config before call this.
+func (s *Scheduler) UpdateSourceCfg(cfg *config.SourceConfig) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if !s.started {
+		return terror.ErrSchedulerNotStarted.Generate()
+	}
+
+	// 1. check whether the config exists.
+	_, ok := s.sourceCfgs[cfg.SourceID]
+	if !ok {
+		return terror.ErrSchedulerSourceCfgNotExist.Generate(cfg.SourceID)
+	}
+	// 2. check if tasks using this configuration are running
+	if tasks := s.GetTaskNameListBySourceName(cfg.SourceID); len(tasks) > 0 {
+		return terror.ErrSchedulerSourceOpTaskExist.Generate(cfg.SourceID, tasks)
+	}
+	// 3. check this cfg is ok to update.
+	if !checkSourceCfgCanUpdated(s.sourceCfgs[cfg.SourceID], cfg) {
+		return terror.ErrSchedulerSourceCfgUpdate.Generate()
+	}
+	// 4. put the config into etcd.
+	_, err := ha.PutSourceCfg(s.etcdCli, cfg)
+	if err != nil {
+		return err
+	}
+	// 5. record the config in the scheduler.
+	s.sourceCfgs[cfg.SourceID] = cfg
+	return nil
+}
+
+// currently the source cfg can only update relay-log related parts.
+func checkSourceCfgCanUpdated(oldCFG, newCFG *config.SourceConfig) bool {
+	newCFGClone := newCFG.Clone()
+	newCFGClone.RelayBinLogName = oldCFG.RelayBinLogName
+	newCFGClone.RelayBinlogGTID = oldCFG.RelayBinlogGTID
+	newCFGClone.RelayDir = oldCFG.RelayDir
+	newCFGClone.Purge = oldCFG.Purge
```

Comment:
seems not,  when enable relay in source_worker, it will init a new relayholder,and it will use the latest cfg  from ETCD, so this change should work🤔

https://github.com/pingcap/dm/blob/8e3fc124a2df9d76e88146facb8a824671b06f8e/dm/worker/source_worker.go#L309

Reply:
what do you want to express about "after the DM-worker restart it will see the new config, and before the restart it has used the old config"? do you mean when the old config is updated, it will init a new relayholder?

Reply:
Yes, that's what I meant, when the source cfg is updated, the relayholder is will also be updated with the new cfg

Reply:
How does the existed relayholder know that the source config is updated?

Reply:
seems there is no existed relayholder? because no relay is enabled on this worker